.travis.yml

```diff
@@ -13,69 +13,140 @@ env:
+  - test
+  - name: e2e-test
+    if: type = pull_request
+  - name: build
+    if: (type != pull_request AND branch = master AND tag IS blank) OR tag IS present
+  - name: deploy-master
+    if: type != pull_request AND branch = master AND tag IS blank
+  - name: release
+    if: tag IS present
+
+jobs:
+  include:
+    #deps cache stage
+    - stage: dep-cache
+      name: "Vendor update"
+      script:
+        - source bin/travis_scripts/ensure_glide.sh $BUILD_TOOLS_PATH "v0.13.1"
+        - glide "-home" $GLIDE_HOME install
+
+    #test stage
+    - stage: test
+      script: bin/check_vet
+      name: "Vet check"
+    
+    - script: bin/builder_run bin/check_lint
+      name: "Linter"
+
+    - script: bin/check_fmt
+      name: "Formating check"
+
+    - script: bin/check_license
+      name: "License check"
+
+    - script: bin/test
+      name: "Unit tests"
+
+    #e2e stage
+    - stage: e2e-test
+      script: bin/test_e2e
+      name: "End-to-end tests"
+
+    #build artifacts
+    - stage: build
+      script:
+      - bin/package_all $BUILD_VERSION linux
+      - bin/s3 cp build/package/mysterium_server_linux_amd64.tar.gz s3://build-artifacts
+      - bin/s3 cp build/package/mysterium_client_linux_amd64.tar.gz s3://build-artifacts
+      - bin/s3 cp build/package/mysterium-client_linux_amd64.deb    s3://build-artifacts
+      - bin/s3 cp build/package/mysterium-node_linux_amd64.deb      s3://build-artifacts
+      name: "Linux packages"
+
+    - script:
+      - bin/package_all $BUILD_VERSION windows
+      - bin/s3 cp build/package/mysterium_server_windows_amd64.zip s3://build-artifacts
+      - bin/s3 cp build/package/mysterium_client_windows_amd64.zip s3://build-artifacts
+      name: "Windows packages"
+
+    - script:
+      - bin/package_all $BUILD_VERSION darwin
+      - bin/s3 cp build/package/mysterium_server_darwin_amd64.tar.gz s3://build-artifacts
+      - bin/s3 cp build/package/mysterium_client_darwin_amd64.tar.gz s3://build-artifacts
+      name: "OSX packages"
+      
+    - script:
+      - bin/server_package_docker
+      - docker save mysterium-node:alpine | gzip > mysterium-node_alpine.tgz
+      - bin/s3 cp mysterium-node_alpine.tgz s3://docker-images
+      name: "Alpine docker image for node"
+
+    - script:
+      - bin/client_package_docker
+      - docker save mysterium-client:alpine | gzip > mysterium-client_alpine.tgz
+      - bin/s3 cp mysterium-client_alpine.tgz s3://docker-images
+      name: "Alpine docker image for client"
+
+    - script:
+      - bin/server_package_docker_ubuntu $BUILD_VERSION
+      - docker save mysterium-node:ubuntu | gzip > mysterium-node_ubuntu.tgz
+      - bin/s3 cp mysterium-node_ubuntu.tgz s3://docker-images
+      name: "Ubuntu docker image for node"
+
+    - script:
+      - bin/client_package_docker_ubuntu $BUILD_VERSION
+      - docker save mysterium-client:ubuntu | gzip > mysterium-client_ubuntu.tgz
+      - bin/s3 cp mysterium-client_ubuntu.tgz s3://docker-images
```

Comment:
The idea was to avoid pushing other images if at least one failed since it could be incompatible. So push to docker registry is done on the next stage.